### PR TITLE
Enforce size limits

### PR DESCRIPTION
**Pull request by gbuenodevsuse**
Addresses #10

### CODE REVIEW

**Review by mbelur — CHANGES_REQUESTED** [review]

File: pkg/client/client.go [hunk 1]
```diff
@@ -301,9 +302,20 @@ func (tc *TelemetryClient) Register() (err error) {
 }
 
 func (tc *TelemetryClient) Generate(telemetry types.TelemetryType, content []byte, tags types.Tags) error {
+	log.Printf("Generated Telemetry:\nName: %q\nTags: %v\nContent: %s\nSize: %v\n",
+		telemetry, tags, content, utils.HumanReadableSize(content))
+
+	// Enforce size limits
+	tl := &telemetrylib.TelemetryDataLimits{}
+	tl.SetTelemetryDataLimits(5242880, 10)
```
[the remark this answers is not in the document]
Comment (gbuenodevsuse) [reply pkg/client/client.go:310]: Yes, we can do that. Btw, I just have an arbitrary value for them right now. What would be a good value for max and min size limits?

File: pkg/utils/utils.go [hunk 2]
```diff
@@ -70,3 +73,49 @@ func DecompressGZIP(compressedData []byte) (decompressedData []byte, err error)
...
+
+// TODO: check if it's worth trying to compress the data prior to compressing it
+// This would save some CPU usage client side
+// TODO: have telemetry data type be passed in as a parameter to further check if we should compress data and which algorithm to use
+func ShouldCompress(data []byte) (resultData []byte, compression *string, err error) {
+	// 'compression' is inserted as a sql.NullString, hence it is returned as a nullable string
+	var nilStr *string = nil
+	var validStr string = "gzip"
+
+	compressedData, err := CompressGZIP(data)
```
[the remark this answers is not in the document]
Comment (gbuenodevsuse) [reply pkg/utils/utils.go:99]: I talked to Fergal about this. It is the plan for future changes. I added a TO DO comment to keep track of it as we move forward. I could add the switch cases to get the code ready for the changes, even though we will only have the default case for now, what do you think?

[the remark this answers is not in the document]
Comment (rtamalin) [reply pkg/utils/utils.go:99]: So there are two compression scenarios we are dealing with here...

The over the wire data compression for the telemetry report submissions which comes into play when the telemetry client sends a telemetry report to the server, and in that case we definitely should be thinking in terms of have a switch statement structure to allow for different options that the client may want to use, though for now we can settle for only implementing the gzip case when submitting telemetry reports...

However there is a second compression scenario that we want to address, which is independent of the over the wire scenario and is purely client side, and has nothing to do with the server side compression support, and that is the compression of telemetry data at rest within the client side data store. In the long run we may choose to support multiple compression techniques, which may be based upon the telemetry type or heuristic rules, but initially gzip should be a good starting point as a balance between CPU cost and compression cost... bzip2 is more CPU expensive for slightly better compression for certain data stream patterns, and xz is much more expensive for better compression gains.

When a telemetry source provides us with a telemetry blob, we will want to store that in the client data store in a compressed format, to minimise the amount of disk space consumed by the telemetry client. When we retrieve that telemetry blob from the client data store to send it over the wire we will decompress it so that it can be included uncompressed in the generated telemetry report, maximising the potential compression gain acheieved for the over the wire compression of the telemetry report in those cases where a telemetry report contains multiple similar data items.

File: pkg/client/client.go [hunk 3]
```diff
@@ -24,6 +25,8 @@ const (
 	CONFIG_PATH     = CONFIG_DIR + "/config.yaml"
 	AUTH_PATH       = CONFIG_DIR + "/auth.json"
 	INSTANCEID_PATH = CONFIG_DIR + "/instanceid"
+	MAX_SIZE        = 5242880
```
Comment (rtamalin) [inline pkg/client/client.go:28]: I would probably make the name here a bit more descriptive...

The first question I would ask when I see this is "What is this the max size for?" so probably a good idea to include a prefix appropriate to the object that it is the max size for... E.g. ITEM_MAX_SIZE etc

**Review by rtamalin — CHANGES_REQUESTED** [review]
This PR should focus on the basic plumbing for enforcing the size limits for telemetry data item submissions.

The changes associated with the compression of the telemetry data items at rest in the client data store should be moved to a new PR, independent of this PR.

File: pkg/utils/utils.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gbuenodevsuse) [reply pkg/utils/utils.go]: Reopened issue #12 to further improve on this.
Also ported the changes regarding this matter to a different [PR](https://github.com/SUSE/telemetry/pull/29).
Updates regarding compression will be moved to the new PR as well.

File: pkg/client/client.go [hunk 4]
```diff
@@ -16,14 +16,17 @@ import (
...
 	//CONFIG_DIR  = "/etc/susetelemetry"
-	CONFIG_DIR      = "/tmp/susetelemetry"
-	CONFIG_PATH     = CONFIG_DIR + "/config.yaml"
-	AUTH_PATH       = CONFIG_DIR + "/auth.json"
-	INSTANCEID_PATH = CONFIG_DIR + "/instanceid"
+	CONFIG_DIR              = "/tmp/susetelemetry"
+	CONFIG_PATH             = CONFIG_DIR + "/config.yaml"
+	AUTH_PATH               = CONFIG_DIR + "/auth.json"
+	INSTANCEID_PATH         = CONFIG_DIR + "/instanceid"
+	TELEMETRY_DATA_MAX_SIZE = 10485760
```
Comment (rtamalin) [inline pkg/client/client.go:28]: I would move these constant settings to pkg/lib/limits.go file, since they are related to the limits themselves.

File: pkg/client/client.go [hunk 5]
```diff
@@ -301,9 +304,20 @@ func (tc *TelemetryClient) Register() (err error) {
 }
 
 func (tc *TelemetryClient) Generate(telemetry types.TelemetryType, content []byte, tags types.Tags) error {
+	log.Printf("Generated Telemetry:\nName: %q\nTags: %v\nContent: %s\nSize: %v\n",
+		telemetry, tags, content, utils.HumanReadableSize(content))
+
+	// Enforce size limits
+	tl := &telemetrylib.TelemetryDataLimits{}
```
Comment (rtamalin) [inline pkg/client/client.go:311]: Instead of setting things up here I would define a NewTelemetryDataLimits() routine which returns a pointer to a TelemetryDataLimits struct which has had the limits already set.

Optionally define an Init() method for the struct pointer that does the initialisation, and have the NewTelemetryDataLimits() routine use the Init() method to do the initialisation.

File: testdata/telemetry/SLE-SERVER-SCCHwInfo/sle12sp5-test.json [hunk 6]
```diff
@@ -11,4 +11,4 @@
     "cloud_provider": "",
     "mem_total": 4096
   }
-}
+}
```
Comment (rtamalin) [inline testdata/telemetry/SLE-SERVER-SCCHwInfo/sle12sp5-test.json:14]: You appear to have an unnecessary change in this file?

**Review by rtamalin — CHANGES_REQUESTED** [review]
The code is generally good, just requesting a restructuring of where some of the code is placed.

File: pkg/client/client.go (hunk 5 above)
Comment (rtamalin) [reply pkg/client/client.go:311]: This should all be in the limits.go file itself, with the lib.NewTelemetryDataLimits() call here to obtain a limits struct pointer.

File: pkg/utils/utils.go [hunk 7]
```diff
@@ -70,3 +71,17 @@ func DecompressGZIP(compressedData []byte) (decompressedData []byte, err error)
 
 	return decompressedData, nil
 }
+
+func HumanReadableSize(data []byte) string {
```
Comment (rtamalin) [inline pkg/utils/utils.go:75]: You probably should add a unit test for this routine just to increase our unit test coverage and validate that it is working as expected.

File: pkg/utils/utils.go [hunk 8]
```diff
@@ -70,3 +71,17 @@ func DecompressGZIP(compressedData []byte) (decompressedData []byte, err error)
...
+	size := len(data)
+	if size < unit {
+		return fmt.Sprintf("%d B", size)
+	}
+	div, exp := int64(unit), 0
+	for n := size / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	return fmt.Sprintf("%.1f %cB", float64(size)/float64(div), "KMGTPE"[exp])
```
Comment (rtamalin) [inline pkg/utils/utils.go:86]: Given that you are doing 2^10 rather than 10^3 (disk manufacturer) multipliers, I would use the unambiguous format, i.e. KiB, MiB, etc.

Also probably better to keep 3 places of precision given the multiplier stepping you are using.

```suggestion
	return fmt.Sprintf("%.3f %ciB", float64(size)/float64(div), "KMGTPE"[exp])
```

File: testdata/telemetry/SLE-SERVER-SCCHwInfo/sle12sp5-test.json (hunk 6 above)
Comment (gbuenodevsuse) [reply testdata/telemetry/SLE-SERVER-SCCHwInfo/sle12sp5-test.json:14]: I think this was because of my VS Code formatter. I used this file to validate my changes, and it ended up formatting it.

File: pkg/client/client.go [hunk 9]
```diff
@@ -301,9 +302,15 @@ func (tc *TelemetryClient) Register() (err error) {
 }
 
 func (tc *TelemetryClient) Generate(telemetry types.TelemetryType, content []byte, tags types.Tags) error {
+	// Enforce size limits
+	_, err := telemetrylib.NewTelemetryDataLimits(content)
```
Comment (rtamalin) [inline pkg/client/client.go:306]: A NewSomething() style routine should conform to the Go new() coding pattern, i.e. you call the NewSomething() routine to option a new object that has minimally initialised, ready to call methods on. The arguments to such a NewSomething() call should be related to minimally initialising the created object, e.g. dimensions for initilising contained objects and structures.

The pattern you have here is more of a top-level module helper function style, which internally might get a new object instance, possibly using a NewSomething() routine, and use it to perform the desired action, without requiring the caller to do any intermediate setup work.

I would rename your NewTelemetryDataLimits() helper to CheckTelemetryDataLimits() and there is no need for it to return the TelemetryDataLimits() pointer any more.

But for future potential use cases it would be useful to implement a NewTelemetryDataLimits() helper that returns a ready to use TelemetryDataLimits object pointer.

File: pkg/lib/limits.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	TELEMETRY_DATA_MIN_SIZE = 10
+)
+
+type TelemetryDataLimits struct {
+	MaxSize uint64
+	MinSize uint64
+}
+
+// func NewTelemetryDataLimits(data []byte) *TelemetryDataLimits {
+func NewTelemetryDataLimits(data []byte) (*TelemetryDataLimits, error) {
```
Comment (rtamalin) [inline pkg/lib/limits.go:20]: See comments above about separating this into two helper functions, namely:
* CheckTelemetryDataLimits([]bytes) error
* NewTelemetryDataLimits() *TelemetryDataLimits

File: testdata/telemetry/SLE-SERVER-SCCHwInfo/sle12sp5-test.json (hunk 6 above)
Comment (rtamalin) [reply testdata/telemetry/SLE-SERVER-SCCHwInfo/sle12sp5-test.json:14]: Just add back in the blank line at the end of the file and save and commit that and it should remove this unnecessary change.

Better not to have unnecessary changes, especially whitespace ones, if possible.

**Review by rtamalin — APPROVED** [review]
LGTM
